sdks/python/apache_beam/io/gcp/spanner.py

```diff
@@ -0,0 +1,503 @@
+  purpose.
+
+  *Option 2: specify a custom expansion service*
+
+  In this option, you startup your own expansion service and provide that as
+  a parameter when using the transforms provided in this module.
+
+  This option requires following pre-requisites before running the Beam
+  pipeline.
+
+  * Startup your own expansion service.
+  * Update your pipeline to provide the expansion service address when
+    initiating Spanner transforms provided in this module.
+
+  Flink Users can use the built-in Expansion Service of the Flink Runner's
+  Job Server. If you start Flink's Job Server, the expansion service will be
+  started on port 8097. For a different address, please set the
+  expansion_service parameter.
+
+  **More information**
+
+  For more information regarding cross-language transforms see:
+  - https://beam.apache.org/roadmap/portability/
+
+  For more information specific to Flink runner see:
+  - https://beam.apache.org/documentation/runners/flink/
+"""
+
+# pytype: skip-file
+
+from __future__ import absolute_import
+
+import typing
+import uuid
+from typing import List
+from typing import NamedTuple
+from typing import Optional
+
+from past.builtins import unicode
+
+from apache_beam import coders
+from apache_beam.transforms.external import BeamJarExpansionService
+from apache_beam.transforms.external import ExternalTransform
+from apache_beam.transforms.external import NamedTupleBasedPayloadBuilder
+from apache_beam.typehints.schemas import named_tuple_to_schema
+
+__all__ = [
+    'WriteToSpanner',
+    'ReadFromSpanner',
+    'MutationCreator',
+    'TimestampBoundMode',
+    'TimeUnit',
+]
+
+
+def default_io_expansion_service():
+  return BeamJarExpansionService(
+      'sdks:java:io:google-cloud-platform:expansion-service:shadowJar')
+
+
+WriteToSpannerSchema = typing.NamedTuple(
+    'WriteToSpannerSchema',
+    [
+        ('instance_id', unicode),
+        ('database_id', unicode),
+        ('project_id', Optional[unicode]),
+        ('max_batch_size_bytes', Optional[int]),
+        ('max_number_mutations', Optional[int]),
+        ('max_number_rows', Optional[int]),
+        ('grouping_factor', Optional[int]),
+        ('host', Optional[unicode]),
+        ('emulator_host', Optional[unicode]),
+        ('commit_deadline', Optional[int]),
+        ('max_cumulative_backoff', Optional[int]),
+    ],
+)
+
+
+class WriteToSpanner(ExternalTransform):
+  """
+  A PTransform which writes mutations to the specified instance's database
+  via Spanner.
+
+  This transform receives Mutations defined as NamedTuple which are created
+  via utility class MutationCreator. Mutation needs to know what row type does
+  it wrap. Example::
+
+    ExampleRow = typing.NamedTuple('ExampleRow',
+                                   [('id', int), ('name', unicode)])
+    coders.registry.register_coder(ExampleRow, coders.RowCoder)
+
+    mutation_creator = MutationCreator('table', ExampleRow, 'ExampleMutation')
```

Comment:
Overall I think it makes a lot of sense to use Rows for the Mutations, with a nested Row for the data, but this API is pretty tricky. Could you look into adding a separate PTransform (or multiple PTransforms) for converting the Rows to mutations? I think an API like this should be possible:

```py
pc = ... #some PCollection with a schema

pc | RowToMutation.insert('table')
     | WriteToSpanner(...)

OR 

pc | RowToMutation.insertOrUpdate('table')
     | WriteToSpanner(...)

OR

pc | RowToMutation.delete('table')
     | WriteToSpanner(...)
```

The PTransform would be able to look at the `element_type` of the input PCollection and create a mutation type that wraps it  in the `expand` method. There's not a lot of examples of logic like this in the Python SDK (yet) the only one I know of is here: https://github.com/apache/beam/blob/cfa448d121297398312d09c531258a72b413488b/sdks/python/apache_beam/dataframe/schemas.py#L50-L55

That way the user wouldn't need to pass the type they're planning on using to MutationCreator. What do you think of that?

Reply:
That way we loose possibility of mixing different kinds of mutations. I don't imagine any sane usage of mixed insert/delete as the order is not guaranteed. Since we will always map rows to mutations before then it would be good to enclose mapping rows to mutations inside WriteToSpanner. How about such an API?:
```
WriteToSpanner(...).insert(table)
WriteToSpanner(...).delete(table)
```
It's not consistent with ReadFromSpanner(...) but I think it's better than forcing the user to call RowToMutation each time.
To be more consistent I could do something like `ReadFromSpanner(...).from_table(table)` and `ReadFromSpanner(...).from_sql(sql_query)`